Server side

yukinorei opened the pull request:
# 実装
犬の写真が表示されるよう修正
テストコード実装

# イメージ
成功
<img width="361" alt="2018-12-06 0 51 32" src="https://user-images.githubusercontent.com/34594829/49525562-63349f80-f8f1-11e8-9177-49318394b152.png">

失敗
<img width="690" alt="2018-12-06 0 51 55" src="https://user-images.githubusercontent.com/34594829/49525574-66c82680-f8f1-11e8-906d-4609561070aa.png">

テスト結果
<img width="486" alt="2018-12-06 0 49 35" src="https://user-images.githubusercontent.com/34594829/49525594-6def3480-f8f1-11e8-9035-897c064ccae1.png">

Mitsuru53 commented on server/utils.js:
```diff
@@ -1,22 +1,14 @@
+  const originalPathArray = originalPath.split('/');
+  const removeColonString = pathWithParams.split(':').join('');
+  const pathWithParamsArray = removeColonString.split('/');
+  pathWithParamsArray.forEach((key, index) => {
```
ここはチェックがもう少し丁寧にしたいですね！
今回はurlが決め打ち('/api/dog/:id')になっているのですが、
これだと、決められた特定のurlしか対応できないですね
今回のappでは、考慮しなくてもいいかもですが、
せっかくなので練習がてら、柔軟にurlに対応できるように修正してみましょう

yukinorei replied:
これは・・・
例えば、
originalPathが'/api/:id/dog'とかでpathWithParamsが'/api/dog/1みたいになってても、
id: 1　と取得できるようにするって理解であってますか？

Mitsuru53 replied:
例えば、
‘/api/dog/1’, ‘/api/dogs/:id’
みたいにurlが一致しない場合を考慮したいって感じですね！
言葉足らずで申し訳なかったです！

yukinorei replied:
なるほど！承知しました！！

clonable-eden commented:
<img width="547" alt="2018-12-13 0 00 33" src="https://user-images.githubusercontent.com/443672/49878073-22490780-fe6a-11e8-96d6-299bedc0cc7b.png">

```sh
$ npm test
```
を実行するとエラーになりますね。`pathWithParamsArray`はその名の通りArrayなので、`startsWith`なんてないよ！って言ってます。おそらく`isParamsValue`の引数として渡した何かしらの文字列に対して`startsWith`したいのかな、と妄想しました。

また、仮にその修正を行ったとしても https://github.com/yukinorei/tpa-04-dog-service/pull/1#discussion_r240150799 こちらは満たせないのではないかな？と思います。`isParamValue`が`false`のものも、パスが一致するかをチェックしないといけないですよね！

yukinorei commented:
testでisParamsValueを使うために、parsePathParametersの外に出すのと、pathWithParamsArray[i]の値のstarts.With(':')となるようにしてstartsWithなんてないよ！ってエラーに関しては解消できたのですが、
それでもうまく犬の画像が出なくなってしまいました。
でもutilsは他に何がおかしいのか考えてもわからないので教えていただけますでしょうか。
utilsは以下のように修正する予定です。

```
const isParamsValue = function(item) {
  return item.startsWith(':');
};

const parsePathParameters = function(originalPath, pathWithParams) {
  const parametersObj = {};
  const originalPathArray = originalPath.split('/');
  const pathWithParamsArray = pathWithParams.split('/');
  for (let i = 1; i < pathWithParamsArray.length; i += 1) {
    if (isParamsValue(pathWithParamsArray[i])) {
      const paramKey = pathWithParamsArray[i].slice(1);
      const paramValue = originalPathArray[i];
      parametersObj[paramKey] = paramValue;
    }
  }
  return parametersObj;
};

module.exports = {
  isParamsValue,
  parsePathParameters,
};
```

Mitsuru53 commented:
> でもutilsは他に何がおかしいのか考えてもわからないので教えていただけますでしょうか。

テストを実行してみました
テストが失敗してますね。こちらが通るようにコードを修正しましょう

<img width="446" alt="2018-12-25 19 45 31" src="https://user-images.githubusercontent.com/23156390/50420950-a72a1d00-087d-11e9-8e8e-d64a2067d41c.png">
